src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,249 @@
+  schemas_.push_back(schema->Copy());
+  exec_queries_.push_back(std::move(exec_query));
+
+  *idx = exec_queries_.size() - 1;
+  return true;
+}
+
+bool QueryExecUtil::ExecuteQuery(size_t idx, TupleFunction tuple_fn,
+                                 common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+                                 common::ManagedPointer<metrics::MetricsManager> metrics,
+                                 const execution::exec::ExecutionSettings &exec_settings) {
+  NOISEPAGE_ASSERT(idx < exec_queries_.size(), "Invalid query index");
+  NOISEPAGE_ASSERT(txn_ != nullptr, "Requires BeginTransaction() or UseTransaction()");
+  auto txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  planner::OutputSchema *schema = schemas_[idx].get();
+  auto consumer = [&tuple_fn, schema](byte *tuples, uint32_t num_tuples, uint32_t tuple_size) {
+    if (tuple_fn != nullptr) {
+      for (uint32_t row = 0; row < num_tuples; row++) {
+        uint32_t curr_offset = 0;
+        std::vector<execution::sql::Val *> vals;
+        for (const auto &col : schema->GetColumns()) {
+          auto alignment = execution::sql::ValUtil::GetSqlAlignment(col.GetType());
+          if (!common::MathUtil::IsAligned(curr_offset, alignment)) {
+            curr_offset = static_cast<uint32_t>(common::MathUtil::AlignTo(curr_offset, alignment));
+          }
+
+          auto *val = reinterpret_cast<execution::sql::Val *>(tuples + row * tuple_size + curr_offset);
+          vals.emplace_back(val);
+          curr_offset += execution::sql::ValUtil::GetSqlSize(col.GetType());
+        }
+
+        tuple_fn(vals);
+      }
+    }
+  };
+
+  // Create ExecutionContext with no metrics to prevent recording
+  execution::exec::OutputCallback callback = consumer;
+  auto accessor = catalog_->GetAccessor(txn, db_oid_, DISABLED);
+  auto exec_ctx = std::make_unique<execution::exec::ExecutionContext>(
+      db_oid_, txn, callback, schema, common::ManagedPointer(accessor), exec_settings, metrics, nullptr);
+
+  exec_ctx->SetParams(common::ManagedPointer<const std::vector<parser::ConstantValueExpression>>(params.Get()));
+
+  exec_queries_[idx]->Run(common::ManagedPointer(exec_ctx), execution::vm::ExecutionMode::Interpret);
+  return true;
+}
+
+bool QueryExecUtil::ExecuteDML(const std::string &query,
+                               common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+                               common::ManagedPointer<std::vector<type::TypeId>> param_types, TupleFunction tuple_fn,
+                               common::ManagedPointer<metrics::MetricsManager> metrics,
+                               std::unique_ptr<optimizer::AbstractCostModel> cost,
+                               const execution::exec::ExecutionSettings &exec_settings) {
+  uint64_t idx = 0;
+  if (!CompileQuery(query, params, param_types, std::move(cost), exec_settings, &idx)) {
```

Comment:
We should actually cache the compiled internal query. This can be local to a `QueryExecUtil`. You just need to change the query vector to a map. I think the query strings should be sufficient to uniquely determine a query.